### ISSUE TITLE
`EventStream` doesn't handle dropped connections

### ISSUE THREAD

**Issue by klaussner** [particle-iot/particle-api-js#95]
The `EventStream` returned from `Particle#getEventStream` doesn't reconnect if the connection is dropped because of network issues. Instead, the stream remains open but doesn't receive any events.

**Reproduction**

You can easily reproduce this issue with the Particle CLI (see particle-iot/particle-cli#256) . Run `particle subscribe` in a terminal, disable WiFi/remove your Ethernet cable, wait for ~2 minutes (depending on the operating system) until the underlying socket is closed on the server, reconnect to the Internet, and run `particle publish e --private` in a second terminal. No event will be logged in the first terminal.

**Details**

I think there are currently two problems that prevent automatic reconnects from working:

1. No timeout is set for the response. If the stream doesn't receive data for a certain period of time, it should assume that the connection is lost and try to reconnect. This wouldn't affect streams that don't receive events during the timeout period because the Particle server sends a heartbeat every few seconds. Something like the following could be added to the [`req.on('response', …)`](https://github.com/particle-iot/particle-api-js/blob/d8a27c2a3cdec959d2934424112a89704c47cd74/src/EventStream.js#L42) callback:
```js
res.socket.setTimeout(10000);

res.socket.on('timeout', () => {
  // This will trigger the 'end' event on the response:
  res.socket.destroy();
});
```

2. The stream assumes that the connection can be re-established [almost immediately after the `'end'` event is triggered](https://github.com/particle-iot/particle-api-js/blob/d8a27c2a3cdec959d2934424112a89704c47cd74/src/EventStream.js#L101-L106). If the first attempt to reconnect after `reconnectInterval` milliseconds fails (likely to be the case if the disconnect is caused by network issues), an error will be emitted and the stream won't try to reconnect again.

**Comment by ricott** [comment]
Any plans to solve this bug or any ideas on how I can work around it myself and check if connection is open or not?

**Comment by khpeterson** [comment]
+1 on this issue.  For me the issue boils down to errors from EventStream.req after the stream has connected/the promise has been resolved.  The call to reject that is registered with req.on('error'...) does nothing because the promise has already been resolved and so the errors fall on the floor.  To work around this I'm re-registering my own EventStream.req error handler in the resolve function but obviously it would cleaner to do this inside EventStream.js.  I will create a pull request.

**Comment by monkbroc** [comment]
The work in progress PR is here if you want to test it: https://github.com/particle-iot/particle-api-js/pull/99

**Comment by khpeterson** [comment]
Thank you for the pointer, I reviewed EventStream.js and it looks good to me / addresses the issue I've been fighting.  I will test it in my environment too.  Any idea on when this version will end up in cdn.jsdelivr.net?

**Comment by monkbroc** [comment]
I'd like to test the changes on console.particle.io and the Particle CLI.
For the version, it looks like a major version bump because if user code
depends on the detailed implementation of the event stream, it might break.

**Comment by monkbroc** [comment]
This is now available in version 8.0.0 of particle-api-js
